src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Serialization/ObjectJsonConverter`1.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using Newtonsoft.Json;
+
+namespace Microsoft.AspNetCore.Razor.ProjectEngineHost.Serialization;
+
+internal abstract class ObjectJsonConverter<T> : JsonConverter<T>
+    where T : class
+{
+    protected abstract T ReadFromProperties(JsonDataReader reader);
+    protected abstract void WriteProperties(JsonDataWriter writer, T value);
+
+    public sealed override T? ReadJson(JsonReader reader, Type objectType, T? existingValue, bool hasExistingValue, JsonSerializer serializer)
+    {
+        if (reader.TokenType == JsonToken.Null)
```

Comment:
❓Is JsonToken.Null used for null value literals, or just if the reader has no data? If the former, do we need to call ReadToken to the reader forward?

Reply:
It is used for null literals.